### DISCUSSION TITLE
Setting a property to a value of item to be uploaded in an array returns InvalidArgumentException: Type is not supported

### DISCUSSION THREAD

**Discussion by faisuc** [livewire/livewire#3122] [7. Bugs (archived)]
### Description

Setting a property to a value of item to be uploaded in an array returns

`InvalidArgumentException: Type is not supported`

Using dot notation works fine.

### Exact steps to reproduce

This will not work

![image](https://user-images.githubusercontent.com/7190009/104544674-4c7ddc80-5663-11eb-9613-6bb27166681d.png)

This will work

![image](https://user-images.githubusercontent.com/7190009/104544743-7800c700-5663-11eb-92cf-cdef78cb56ab.png)

**Context**
- Livewire version: 2.3
- Laravel version: 8.12

**Comment by itepifanio** [comment]
Same error when I'm trying to emit a event with a fake file.
- PHP 7.4
- Laravel 8.12 
- Livewire 2.3

**Comment by jartaud** [comment]
Same error:

```
$image = UploadedFile::fake()->image('image.png');
$pdf = UploadedFile::fake()->create('pdf.pdf');


Livewire::test(DocumentCreate::class)
    ->set('documents', [
        ['description' => 'Description 1', 'doc_path' => $image, ],
        ['description' => 'Description 2', 'doc_path' => $pdf, ],
    ])
   ...
```

I get my tests passed using dot notation as @faisuc started:

```
Livewire::test(DocumentCreate::class)
            ->set('documents.0.description', 'Description 1')
            ->set('documents.0.doc_path', $image)
            ->set('documents.1.description', 'Description 2')
            ->set('documents.1.doc_path', $pdf)
            ->call('create')
```

Livewire 2.3.8
Laravel 8.25.0

**Comment by joshhanley** [comment]
@faisuc and @jartaud I don't think it should work that way, the reason is, the test actions like `set` are meant to simulate input from the user/browser. You wouldn't input a whole array into an input field in your browser, you would fill in each of the fields manually. Hence why you need to use dot notation and set them one at a time.

Can you explain more why you would need to pass an array into set?

If you want to pre-populate data in the component, you can pass an array as the second parameter into the `test` function with the keys being the public property name you want to populate. Like

```php
Livewire::test(DocumentCreate::class, ['documents' => [
        ['description' => 'Description 1', 'doc_path' => $image, ],
        ['description' => 'Description 2', 'doc_path' => $pdf, ],
    ])
```

Hope this helps!

**Comment by jartaud** [comment]
@joshhanley I think the problem is that stream resource (the file) is not serializable to JSON. 

![Screenshot from 2021-02-07 01-40-56](https://user-images.githubusercontent.com/266818/107138960-d5e0a180-68e5-11eb-9674-dd423ab8265a.png)


```
// DocumentCreate::class

public array $documents = [];

public function addDocument()
{
    $this->documents[] = ['description' => '', 'doc_path' => ''];
}

public function removeDocument($index)
{
    unset($this->documents[$index]);
}

public function create()
{
    $this->validate([
        'patient' => ['required', 'numeric', 'min:1',
            Rule::in($this->tenant->clients->pluck('id')->all())
        ],
        'documents' => ['array'],
        'documents.*.description' => ['required', 'between:3,255', 'distinct', ],
        'documents.*.doc_path' => ['required', 'file', 'mimes:jpg,jpeg,png,pdf,doc,docx,ods,odt,odp', 'max:2048', ],
    ]);

    foreach ($this->documents as $document) {
        Document::create([
            'tenant_id' => $this->tenant->id,
            'client_id' => $this->patient,
            'description' => $document['description'],
            'doc_path' => $this->storedDocs[] = $document['doc_path']->store("documents/{$this->tenant->id}", 'public'),
        ]);
    }
}
```
```
blade
@foreach ($documents as $key => $item)
<x-input type="text" name="description" wire:model.defer="documents.{{ $key }}.description" required />
<x-input type="file" name="doc_path" wire:model.defer="documents.{{ $key }}.doc_path" required />
@endforeach
```

**Livewire will produce the payload based on:**
```
<input type="text" name="documents[0][description]" />
<input type="file" name="documents[0][doc_path]" />

<input type="text" name="documents[1][description]" />
<input type="file" name="documents[1][doc_path]" />
```
